platform/email/platform-email-impl/src/main/java/org/codice/ddf/platform/email/impl/SmtpClientImpl.java

```diff
@@ -126,15 +127,34 @@ public Session createSession() {
     properties.setProperty(SMTP_HOST_PROPERTY, hostName);
     properties.setProperty(SMTP_PORT_PROPERTY, portNumber.toString());
 
+    // Jakarta Mail uses java.util.ServiceLoader.load() to look for service implementations, in
+    // this case of the javax.mail.Provider interface. ServiceLoader relies on the Thread Context
+    // Classloader which is not necessarily the correct behavior in an OSGi context and may result
+    // in services not being found.
+    //
+    // To fix this, we set the current thread classloader to the classloader of the bundle
+    // containing those services (making sure to restore the original classloader before returning)
+    // so that ServiceLoader will always find them.
+    //
+    // TODO: Look into SPI Fly. Might be a better solution
+    // http://aries.apache.org/modules/spi-fly.html
+    ClassLoader originalContextClassLoader = Thread.currentThread().getContextClassLoader();
+    Thread.currentThread().setContextClassLoader(Provider.class.getClassLoader());
+
+    Session session;
     if (StringUtils.isNotBlank(userName)) {
       properties.put(SMTP_AUTH_PROPERTY, TRUE);
       properties.put(SMTP_START_TLS_ENABLE_PROPERTY, TRUE);
 
-      return Session.getInstance(properties, createAuthenticator());
+      session = Session.getInstance(properties, createAuthenticator());
     } else {
       properties.setProperty(SMTP_AUTH_PROPERTY, FALSE);
-      return Session.getInstance(properties);
+
+      session = Session.getInstance(properties);
     }
+
+    Thread.currentThread().setContextClassLoader(originalContextClassLoader);
```

Comment:
✏️ Would it be beneficial to reset the class loader in a finally block?

Reply:
Good idea

Reply:
✏️ Should the `Thread.currentThread().setContextClassLoader(Provider.class.getClassLoader());` line go inside the try block in case something goes wrong between that line and when the try block starts?

Reply:
Ah, yeah. I changed it